src/query/api/v1/handler/influxdb/write.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+	xtime "github.com/m3db/m3/src/x/time"
+	"go.uber.org/zap"
+)
+
+const (
+	InfluxWriteURL = handler.RoutePrefixV1 + "/influxdb/write"
+)
+
+type ingestWriteHandler struct {
+	downsamplerAndWriter ingest.DownsamplerAndWriter
+	instrumentOpts       instrument.Options
+	promRewriter         *promRewriter
+}
+
+type ingestField struct {
+	name  []byte // to be stored in __name__; rest of tags stay constant for the Point
+	value float64
+}
+
+type ingestIterator struct {
+	// what is being iterated
+	points       []imodels.Point
+	promRewriter *promRewriter
+	pointIndex   int
+	err          xerrors.MultiError
+
+	// following entries are within current point, and initialized
+	// when we go to the first entry in the current point
+	fields         []*ingestField
+	nextFieldIndex int
+	tags           models.Tags
+}
+
+func (self *ingestIterator) populateFields() bool {
+	point := self.points[self.pointIndex]
+	it := point.FieldIterator()
+	n := 0
+	for it.Next() {
+		n += 1
+	}
+	if n == 0 {
+		return false
+	}
+	it.Reset()
+	self.fields = make([]*ingestField, 0, n)
+	for it.Next() {
+		var value float64 = 0
+		switch it.Type() {
+		case imodels.Boolean:
+			v, err := it.BooleanValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding boolean: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			if v {
+				value = 1.0
+			}
+		case imodels.Integer:
+			v, err := it.IntegerValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding integer: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Unsigned:
+			v, err := it.UnsignedValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding unsigned: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = float64(v)
+		case imodels.Float:
+			v, err := it.FloatValue()
+			if err != nil {
+				err = fmt.Errorf("Error decoding float: %w", err)
+				self.err = self.err.Add(err)
+				continue
+			}
+			value = v
+		default:
+			// TBD if we should stick strings as
+			// tags or not; to prevent cardinality
+			// explosion, we drop them for now
+			continue
+		}
+		nlen := len(point.Name())
+		name := make([]byte, nlen+1+len(it.FieldKey()))
+		copy(name, point.Name())
+		copy(name[nlen:], []byte("_"))
+		copy(name[nlen+1:], it.FieldKey())
+
+		name = self.promRewriter.Metric.ToValid(name)
+		self.fields = append(self.fields, &ingestField{name: name, value: value})
+	}
+	return true
+}
+
+func (self *ingestIterator) Next() bool {
+	for len(self.points) > self.pointIndex {
+		if self.nextFieldIndex == 0 {
+			// Populate tags only if we have fields we care about
+			if self.populateFields() {
+				point := self.points[self.pointIndex]
+				ptags := point.Tags()
+				tags := models.NewTags(len(ptags), nil)
+				for _, tag := range ptags {
+					name := self.promRewriter.Label.ToValid(tag.Key)
```

Comment:
Hm... think that there may be a chance of `name` clashing across different tags which can lead to undefined behavior (i.e. if you have two labels, `foo.bar` and `foo_bar`, they both evaluate to `foo_bar`)

Reply:
This one is the only PR comment I am unsure about; isn't it possible to have similar thing occur also in other ingestion endpoints? Preventing it seems expensive, and as they may come in separate requests, it is not clear to me how it is even doable for the general case (unless maintaining some global mapping which guarantees uniqueness of the rewrite results).

Reply:
Correct. Also sticking in __name__ (whether intentionally or not) seems like bad idea, so I will prevent that too.